### PR TITLE
TandemFoil: 4L/256d architecture transfer at lr=2e-4

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The 4L/256d architecture delivered a -22.3% improvement on AirfRANS (PR #2727). TandemFoil has been running on the default 3L/192d configuration. We expect the same architecture scaling benefit to transfer: more layers and wider hidden dim should improve surface pressure MAE on TandemFoil, particularly with the already-established optimal lr=2e-4 for this dataset.

## Instructions
Run the baseline TandemFoil command with the 4L/256d architecture changes:

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 2e-4 \
  --cosine-t-max 10 \
  --no-use-ema \
  --model-slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --wandb_group tandem-arch-scaling \
  --epochs 9999
```

Changes vs baseline:
- `--model-layers 4` (was 3)
- `--model-hidden-dim 256` (was 192)
- `--model-heads 4` (was default)

## Baseline
Current best TandemFoil metrics (PR #2610, W&B: ixs1rqgk):
- val_primary/surface_pressure_mae: **45.07** (lower is better)

Reproduce baseline:
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT sasuke:

## Results

### TandemFoil: 4L/256d architecture transfer at lr=2e-4

| Metric | Baseline (3L/192d) | 4L/256d | Δ |
|--------|-------------------|---------|---|
| **val_primary/surface_pressure_mae** | **45.07** (ep 107) | **46.32** (ep 96) | **+2.8% (worse)** |
| test_primary/surface_pressure_mae | — | 52.62 | — |
| test_single_in_dist | — | 55.18 | — |
| test_geom_camber_rc | — | 58.94 | — |
| test_geom_camber_cruise | — | 50.05 | — |
| test_re_rand | — | 46.30 | — |

- **W&B run:** b40w88kd
- **Epochs:** 104 (180-min timeout, ~1.7 min/epoch vs ~1.5 min/epoch for 3L/192d)
- **Peak grad norm mean:** 6.46 (epoch 1) → 0.55 (epoch 104), stable descent
- **Final train loss:** 0.103

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema \
  --model-slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --model-layers 4 --model-hidden-dim 256 \
  --model-heads 4 --wandb-group tandem-arch-scaling --agent sasuke \
  --wandb-name "sasuke/tandem-4L256d-lr2e4" --epochs 9999
```

### What happened

**The 4L/256d architecture did NOT beat the 3L/192d baseline on TandemFoil.** Best val trough was 46.32 at epoch 96 vs baseline 45.07 at epoch 107 — a +2.8% regression.

Key observations:
1. **Slower epochs:** 4L/256d runs ~1.7 min/epoch vs ~1.5 min/epoch for 3L/192d, resulting in 104 epochs vs 119 in the same 180-min budget — 13% fewer training steps.
2. **Similar convergence pattern:** Both models show cosine cycling oscillation with T_max=10. The trough envelope descends similarly but 4L/256d doesn't find a deeper basin.
3. **Throughput vs capacity tradeoff:** Unlike AirfRANS (where 4L/256d won by -22.3%), TandemFoil at lr=2e-4 appears to be in a regime where more training steps at smaller model size outperforms fewer steps at larger size. This mirrors the DrivAerML finding where 4L/320d beat 4L/384d due to throughput advantage.
4. **Trough still descending:** Best at epoch 96 of 104 — the model hadn't fully converged, but the gap to baseline is small (1.25 MAE points) and unlikely to close given the epoch deficit.
5. **Note:** `test_geom_camber_cruise/mae_vol_p` showed Infinity — likely a numerical issue in volume pressure at that test split.

### Suggested follow-ups
1. **5L/256d at lr=2e-4** — PR #2595 showed 5L/256d at lr=3e-4 got 52.81. Combining 5L/256d with the better lr=2e-4 could beat the 3L/192d baseline if the capacity boost is large enough to offset the throughput loss.
2. **3L/256d at lr=2e-4** — Test whether width scaling alone (without depth) helps at this LR. Faster epochs than 4L/256d while being wider than 3L/192d.
3. **4L/256d with grad-clip** — Gradient clipping helped dramatically on AirfRANS with 4L/256d. May stabilize the larger model's training on TandemFoil too.
4. **Longer training at 3L/192d** — The current best was still improving at 119 epochs. A 360-min or larger budget could push the baseline much lower.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2772 because ADVISOR: Closing — 4L/256d at lr=2e-4 achieves 46.325 vs baseline 44.72 (3.6% worse). 4L/256d is too deep for TandemFoil: the dataset benefits from 3L/192d's capacity-efficiency tradeoff. The baseline 3L/192d at lr=1.5e-4 has 115 training epochs; 4L/256d at the same LR is slower (2.7 min/ep vs 1.5 min/ep), gets fewer epochs, and the added depth destabilizes late training. Consistent with violet's earlier 5L/256d failure. TandemFoil sweet spot is 3L/192d. Assigning you a new experiment..
